fix: added checks for sysvars and programs (ACKEE-4)

ICavlek opened the pull request:
**Referenced issue**

Resolves ACKEE-4 issue which corresponds to additional checks for sysvars and programs.

**Summary of changes**

- [ ] Does this PR change behavior or adds a new feature ?
- [ ] docs updated
- [ ] tests updated
- [ ] code style matches guidelines

**Reviewer recommendations**

- Check added validations
- Check added unit tests
- Check bindings tests
- Check added accounts to some bindings

**Ready-ready checklist**

- [ ] Labels (ticket type and component) are set
- [ ] Project & Phase is assigned
- [ ] This checklist is removed

ICavlek committed:
chore: added program and sysvar checks in its

ICavlek committed:
test: unit tests for programs and sysvar checks

frenzox commented on programs/axelar-solana-its/src/processor/mod.rs:
```diff
@@ -552,14 +567,13 @@ fn process_set_pause_status(accounts: &[AccountInfo<'_>], paused: bool) -> Progr
+    accounts: &'a [AccountInfo<'a>],
+    chain_name: String,
+) -> ProgramResult {
+    let (payer, program_data_account, its_root_pda, system_account) = extract_accounts(accounts)?;
```
Perhaps update the name to be consistent with the role management one? Or at least to be a bit more specific since I don't think this function can be used for the other instructions

```suggestion
    let (payer, program_data_account, its_root_pda, system_account) = get_trusted_chain_accounts(accounts)?;
```

ICavlek replied:
Done

frenzox commented:
Looks good so far!

frenzox commented on bindings/prepare_test.sh:
Should we perhaps have something in the CI to run this and the binding tests? I just realized I broke this with the flattening of the directory structure and didn't even realize ><

ICavlek replied:
No worries, @eloylp and I already talked about adding it to CI and updating `native-to-anchor` so that we don't have to manually re-adjust things. Due to other critical things, we have decided to put the whole bindings restructuring part aside for now.

ICavlek committed:
chore: proper function naming

ICavlek committed:
chore: memo test fix

ICavlek committed:
chore: updated validation for GiveTokenAccounts

ICavlek committed:
chore: updated validation for TakeTokenAccounts

ICavlek committed:
chore: clippy

ICavlek committed:
chore: refactor RoleManagementAccounts

ICavlek committed:
chore: validate system accounts in the rest of programs

ICavlek committed:
test: refactor boilerplate test code

ICavlek committed:
feat: added Validate trait

ICavlek committed:
feat: check of system account

ICavlek committed:
chore: refactor of system account check

ICavlek committed:
bugfix: inproper validation

ICavlek committed:
feat: updated FromAccountInfoSlice trait

ICavlek committed:
chore: refactor functions for validation

ICavlek committed:
chore: remove unnecessary tests

ICavlek committed:
chore: remove additional unnecessary test

ICavlek committed:
chore: SetFlowLimitAccounts as other structs

ICavlek committed:
chore: clippy fix

eloylp commented on bindings/generated/axelar-solana-its/src/program.ts:
```diff
@@ -330,6 +330,11 @@ type AxelarSolanaIts = {
+          name: "gatewayRootPda";
+          isMut: false;
+          isSigner: false;
+        },
```
Suprised about this change ! Do we maybe removed more than the account on previous pR :thinking:

ICavlek replied:
I really don't know if it was removed or other thing. It took me a while to find this one, as I was verifying passed accounts from bindings :disappointed:

eloylp commented on programs/axelar-solana-its/src/processor/link_token.rs:
```diff
@@ -209,7 +209,6 @@ pub(crate) fn register_custom_token<'a>(
 pub(crate) fn register_canonical_interchain_token<'a>(
     accounts: &'a [AccountInfo<'a>],
 ) -> ProgramResult {
-    msg!("Instruction: RegisterCanonicalInterchainToken");
```
Why ? :)

ICavlek replied:
This `msg!()` was only used as a checkpoint in bindings tests. I have updated `RegisterCanonicalInterchainToken` bindings test to verify data on even deeper `msg!("Instruction: Register Token")`.
Therefore, this one is unnecessary.

ICavlek commented:
Ready for review:

- Implemented each program validation in separate function for easier unit testing and implementation
- Added `validate_system_account_key` in significant amount of functions
- Added trait `Validate` for structs to call `validate` and use program validation functions
- Extended trait `FromAccountInfoSlice` to must use `Validate` trait; set `from_account_info_slice` as default function
- Implemented `Validate` trait for structs `GmpAccounts`, `DeployInterchainTokenAccounts`, `TakeTokenAccounts`, `GiveTokenAccounts`, `AxelarInterchainTokenExecutableAccounts`, `DeployTokenManagerAccounts` and `SetFlowLimitAccounts` 
- Refactored `SetFlowLimitAccounts` to be consistent with other structs
- `RoleManagementAccounts` with only single `validate_system_account_key`, not with `Validate` because it is located in `helpers` crate and not `programs`
- Updated bindings tests

eloylp reviewed (commented):
Looks good, just a first round of doubts :) - i am continuing reviewing.

frenzox reviewed (approved):
lgtm :rocket:
